aws/resource_aws_pinpoint_app.go

```diff
@@ -0,0 +1,435 @@
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"start": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsPinpointAppCreate(d *schema.ResourceData, meta interface{}) error {
+	pinpointconn := meta.(*AWSClient).pinpointconn
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	} else if v, ok := d.GetOk("name_prefix"); ok {
+		name = resource.PrefixedUniqueId(v.(string))
+	} else {
+		name = resource.UniqueId()
+	}
+
+	log.Printf("[DEBUG] Pinpoint create app: %s", name)
+
+	req := &pinpoint.CreateAppInput{
+		CreateApplicationRequest: &pinpoint.CreateApplicationRequest{
+			Name: aws.String(name),
+		},
+	}
+
+	output, err := pinpointconn.CreateApp(req)
+	if err != nil {
+		return fmt.Errorf("[ERROR] creating Pinpoint app: %s", err)
+	}
+
+	d.SetId(*output.ApplicationResponse.Id)
+
+	return resourceAwsPinpointAppUpdate(d, meta)
+}
+
+func resourceAwsPinpointAppUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	appSettings := &pinpoint.WriteApplicationSettingsRequest{}
+
+	//if d.HasChange("cloudwatch_metrics_enabled") {
+	//	appSettings.CloudWatchMetricsEnabled = aws.Bool(d.Get("cloudwatch_metrics_enabled").(bool));
+	//}
+
+	if d.HasChange("campaign_hook") {
+		appSettings.CampaignHook = expandCampaignHook(d)
+	}
+
+	if d.HasChange("limits") {
+		appSettings.Limits = expandCampaignLimits(d.Get("limits").(*schema.Set))
+	}
+
+	if d.HasChange("quiet_time") {
+		appSettings.QuietTime = expandQuietTime(d.Get("quiet_time").(*schema.Set))
+	}
+
+	req := pinpoint.UpdateApplicationSettingsInput{
+		ApplicationId:                   aws.String(d.Id()),
+		WriteApplicationSettingsRequest: appSettings,
+	}
+
+	_, err := conn.UpdateApplicationSettings(&req)
+	if err != nil {
+		return err
+	}
+
+	return resourceAwsPinpointAppRead(d, meta)
+}
+
+func resourceAwsPinpointAppRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	log.Printf("[INFO] Reading Pinpoint App Attributes for %s", d.Id())
+
+	app, err := conn.GetApp(&pinpoint.GetAppInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Pinpoint App (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	settings, err := conn.GetApplicationSettings(&pinpoint.GetApplicationSettingsInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Pinpoint App (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	d.Set("name", app.ApplicationResponse.Name)
+	d.Set("application_id", app.ApplicationResponse.Id)
+
+	d.Set("campaign_hook", flattenCampaignHook(settings.ApplicationSettingsResource.CampaignHook))
```

Comment:
We should error check `d.Set()` when attempting to set more complex types like lists and maps to ensure they are properly being read into the Terraform state:

```go
if err := d.Set("campaign_hook", flattenCampaignHook(settings.ApplicationSettingsResource.CampaignHook)); err != nil {
  return fmt.Errorf("error setting campaign_hook: %s", err)
}
if err := d.Set("limits", flattenCampaignLimits(settings.ApplicationSettingsResource.Limits)); err != nil {
  return fmt.Errorf("error setting limits: %s", err)
}
if err := d.Set("quiet_time", flattenQuietTime(settings.ApplicationSettingsResource.QuietTime)); err != nil {
  return fmt.Errorf("error setting campaign_hook: %s", err)
}
```

Reply:
Fixed in https://github.com/terraform-providers/terraform-provider-aws/pull/5956/commits/1ea7cdae069f31c5179a80f274ba1282ba06785d